179 Привязка экрана ленты новостей к апи

TanyaRul opened the pull request:
Не проходит проверку detekt по одному правилу. Указывает на private fun isConnected() в файле RetrofitNetwirkClient: функция имеет 4 оператора возврата, что превышает ограничение в 2. Как лучше сделать: оставить эту функцию и поправить правило или убрать её?

TanyaRul committed:
- connects the news feed screen with the API
- changes the news model
- removes the avatar field from the author object

TanyaRul committed:
- fix detekt

TanyaRul committed:
Merge branch 'develop' into 179-connect-feed-with-api

# Conflicts:
#	app/src/main/java/com/innoprog/android/feature/feed/newsdetails/data/NewsDetailsRepositoryImpl.kt

TanyaRul committed:
- fix detekt

GoetzDeBouville commented on app/src/main/java/com/innoprog/android/feature/feed/newsfeed/data/converters/NewsDtoMapper.kt:
```diff
@@ -0,0 +1,51 @@
+import com.innoprog.android.feature.feed.newsfeed.domain.models.News
+
+class NewsDtoMapper {
+    fun newsDtoToNews(newsDto: NewsDto): News {
```
Лучше сделать в виде методов расширений, так получится изолировать domain от data слоя
Таким образом и не понадобится производить создание объекта NewsDtoMapper, а просто конвертировать возврат из метода репозитория непосредственно в интеракторе, интерфейс интерактора при этом менять не нужно, т.к. через конвертер будут идти данные моделей из domain

TanyaRul replied:
Изменила

GoetzDeBouville commented on app/src/main/java/com/innoprog/android/feature/feed/newsfeed/data/network/RetrofitNetworkClient.kt:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    private fun isConnected(): Boolean {
```
предлагаю написать проверку соединения в виде. функции расширения Context, и использовать уже во всех фичах. 
Пример реализации:
```kotlin
fun Context.checkInternetReachability(): Boolean {
    val connectivityManager = this.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    val activeNetwork =
        connectivityManager.activeNetwork?.let { connectivityManager.getNetworkCapabilities(it) } ?: return false
    return when {
        activeNetwork.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> true
        activeNetwork.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> true
        else -> false
    }
}
```
На такую функцию detekt ругаться не будет, т.к. return уже один

GoetzDeBouville replied:
Не обратил внимание что там два ретурна. Вот вариант с одним возвратом
```kotlin
fun Context.checkInternetReachability(): Boolean {
    val connectivityManager =
        this.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    val activeNetwork = connectivityManager.activeNetwork?.let {
        connectivityManager.getNetworkCapabilities(it)
    }

    return activeNetwork?.let {
        it.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
        it.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) ||
        it.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)
    } ?: false
}

```

TanyaRul replied:
Изменила, но пока оставила в этом же файле. Перепишу на использование общей функции после того, как сольём твой ПР, где она уже вынесена

GoetzDeBouville replied:
Боюсь, это будет не скоро 😂 слишком долго он висит, но уже один чек появился

GoetzDeBouville reviewed (approved):
Оставил пару комментариев

GoetzDeBouville commented on app/src/main/java/com/innoprog/android/feature/feed/newsfeed/data/network/RetrofitNetworkClient.kt:
```diff
@@ -0,0 +1,51 @@
+    NetworkClient {
+
+    override suspend fun loadNewsFeed(): Response {
+        if (!isConnected()) {
```
тут можно воспользоваться функцией not
```
if (isConnected().not())
```
читается лучше чем восклицательный знак и не противоречит правилу [об отрицательных условиях](https://alexlumper.dev/blog/evristicheskie-pravila-i-zapahi-koda#g29-izbegajte-otricatelnyh-uslovij)
Вообще лучше полностью избегать отрицательных условий

TanyaRul committed:
- changes NewsDto mapping
- changes the method of checking the availability of the Internet

TanyaRul committed:
- изменяет модели News, NewsDto, NewsEntity
- добавляет модели Project, ProjectDto
- добавляет методы для получения деталей проекта и отображения их в каждой публикации
- изменяет NewsAdapter, NewsViewHolder
- изменяет метод получения ссылки на изображение в публикации

TanyaRul committed:
- добавляет ресурс для ошибки загрузки информации о проекте

GoetzDeBouville commented on app/src/main/java/com/innoprog/android/feature/feed/newsfeed/data/network/NewsDto.kt:
```diff
@@ -0,0 +1,97 @@
+import com.innoprog.android.feature.feed.newsfeed.domain.models.News
+
+data class NewsDto(
+    @SerializedName("id")
```
Кажется, @SerializedName не обязательно устанавливать для полей с совпадающими именами

Secur1ty-21 replied:
Обязательно, т.к. при обфускации затираются имена полей и gsonConverter уже не может понять по полю

GoetzDeBouville commented on app/src/main/java/com/innoprog/android/feature/feed/newsfeed/data/network/NewsDto.kt:
```diff
@@ -0,0 +1,97 @@
+    val type: String
+)
+
+fun NewsDto.mapToNews(): News {
```
у тебя есть папка converters, наверное лучше там хранить все мапперы

GoetzDeBouville reviewed (approved):
Аппрув, проверил, работает.
Есть пара моментов, не очень понравилось решение провайда viewmodel в adapter и viewholder соответственно, это нарушение принципа единой ответственности, все-таки адаптер это паттерн не про получение данных.
В коде фрагмента увидел companion object со строками, которые стоит вынести все-таки в ресурсы, можно красиво сделать, собрать строки и положить в ресурс типа string-array
Для recyclerview в FeedFragment напрашивается diffutil, т.к. чипы и поиск подразумевают изменения списка (фильтрация элементов)

TanyaRul committed:
- переносит методы маппинга в отдельные файлы
- добавляет строковые ресурсы для использования в чипах
- изменяет принцип заполнения данных проекта в публикацию

TanyaRul committed:
Merge branch 'develop' into 179-connect-feed-with-api

# Conflicts:
#	app/src/main/res/values/strings.xml

Commit:
Remove unnecessary set clickListener in bind method, add BuildConfig feature, small refactoring in viewHolder, add exception messages, add catch exceptions in mapping to domain model

Secur1ty-21 reviewed (commented):
Добавил BuildConfig и логи в catch блоках, убрал случайно оставленный clickListener в bind методе, добавил обработку исключений на маппинг project модели (Если данные с сервера придут битые для одного проекта, то лучше обработать и не отобразить только одну вкладку проекта).  Критических замечаний нет, вынос логики в repository сделан абсолютно правильно

Secur1ty-21 committed:
Update check_build.yml
